Propose feature for troubleshooting running pods

verb opened the pull request:
**What this PR does / why we need it**: Initially described in #27140, this PR proposes a feature to better support minimal container images (i.e. images built from scratch) by adding a new `kubectl debug` to run a container in a pod namespace or create an ad-hoc copy of a pod. This is useful for developers writing native Kubernetes applications that prefer not to ship an entire OS image.

**Special notes for your reviewer**: SIG Node asked me to send this PR to facilitate further discussion.

**Release note**: None, this PR contains only a proposal.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35584)

<!-- Reviewable:end -->

verb committed:
Propose feature to execute ephemeral image in pod

bgrant0607 commented:
cc @kubernetes/sig-node

vishh commented:
@k8s-bot ok to test

liggitt commented on docs/proposals/image-exec.md:
```diff
@@ -0,0 +1,278 @@
+* Is this compatible with Admission control plugins that restrict what images
+  can run on a cluster?
+* Would it be possible to Image Exec into pods that are in a terminal state to
+  determine why they've failed?
```
what security context is used by the new container? how does that interact with PodSecurityPolicy?

liggitt replied:
cc @kubernetes/sig-auth

timstclair replied:
From the sig-node meeting:
- What happens if the connection is dropped?

verb replied:
I looked into this a little today. The current behavior in exec is that the process continues to run but there doesn't seem to be any way to reattach to it. It doesn't get a signal and stdin is left open. My prototype for image exec behaves the same.

A debug container would behave the same, except perhaps that it could be attached with `kubectl attach`, with the kubelet removing the container when the pod is destroyed or when the process exits (potentially relying on the runtime to support automatically removing the container a la "docker run --rm")

liggitt commented on docs/proposals/image-exec.md:
```diff
@@ -0,0 +1,278 @@
+while the latter would provide binaries for kubectl exec.
+
+Adding a container to a running pod would be a large change to Kubernetes, and
+there are several edge cases to consider such as resource limits and monitoring.
```
and admission that locks security context at pod creation time

dchen1107 commented:
Good idea, but need to iron out many details. Here are the several ones from the top of my head:

1) How this container interactive with the pod lifecycle
2) How this debug container present to the end user? show it as another container or hidden as today's podInfraContainer? 
3) resource management introduced by this extra consumption, charging to the existing pod? or update the pod cgroup with extra overhead?

derekwaynecarr commented:
Per my comments on sig-node meeting:

1.  The proposal as defined allows users to launch new containers in a pod that are not discoverable via consumers of the API.  This is problematic for us as we have built security solutions that audit the cluster based on the set of containers/images that are seen running via the API.  We should not allow containers to be run that are associated with a pod that are not discoverable when getting the status of the pod.
2.  I would prefer we explore something similar to init containers by allowing some type of modification to the pod spec for these ephemeral containers that are not part of the core pod lifecycle.
3.  I have some concerns about how SELinux and other tools would work @pmorie

smarterclayton commented:
I agree that this requires discoverability by the API.  The status of this container is important.

A few things:

1. Can we start this container in a terminated (success, failed, or error) pod?  There are many use cases for debugging that, but I also don't want to have to keep the pod's volumes around forever
2. Should this container be allowed to change the pod volumes?  I don't think so - so it may limit some debugging scenarios. 
  1. EXCEPTION: It should be possible to bind mount the other container's filesystems in without changing the pod volumes
3. If we had this as transient-container+attach, there's less of a need for exec (because you could start bash and attach) in many cases people are abusing it.
4. Based on statements we have made before, I think it is inappropriate for the resource limits of the pod to change as a result of this, but that makes it very hard to run one of these inside of a pod that is at its maximum resources (main process has used all available pod memory, for instance).  Does that require pre-reservation?

verb commented:
Summarizing feedback from SIG Node meetup:

@derekwaynecarr added a firm requirement for ability to introspect container images

The following questions were addressed:

1. **Will this provide a view into a container's mount namespace or simply container filesystems?** Only container filesystems since there's no way to support mounting the actual mount namespace as a subdirectory (if this becomes possible we should support it).
2. **How should resources be accounted?** There was weak consensus that it's reasonable to expect users to include resources consumed by an `kubectl exec` in the initial pod sizing.
3. **Dangling pod problem with `kubectl run --rm`?** This would be different from kubectl run in that it would be implemented server-side. That is, the kubelet must ensure there are no dangling containers.

The following open questions were added:

1. How does this affect the pod lifecycle?
2. What are the edge cases of current `kubectl exec`. Can you re-attach to a running `exec`? How would this be implemented for an image exec?
3. Userns + host bindmounts, depending on how userns is implemented, could be messy (hostmount implies no userns, but userns relabeling might be enabled for the whole pod)
4. How does this interact with SELinux?

An alternative proposed during the meeting was to allowing adding a container to the podspec. This would be a larger change, but perhaps also more worthwhile. It could be simplified by introducing a new type of ephemeral container similar to the implementation of init-containers. We ran out of time before discussing whether this solution could meet the other requirements such as container filesystem access.

It's not clear where to go next, so I'm going to work on addressing some of the questions raised by the feedback and try to estimate how much work would be required to support modifying the podspec of a running pod.

smarterclayton commented:
We've discussed the possibility of new container types before and have been
cautious about them.  I think drawing up the use cases and demonstrating
that they are more difficult to reason about if implemented differently is
key.  We'd probably want to find supporting use cases, like the folks at
Huawei and Red Hat who wanted to be able to commit containers in pods (so
they could have a build system for images that uses kube to manage the
containers).  We'd need to rationalize why exec (which is heavily used
today for pod support tasks like backup, debugging, builds, etc) is
different than these containers.

On Tue, Nov 1, 2016 at 5:44 PM, Lee Verberne <notifications@github.com>
wrote:

> Summarizing feedback from SIG Node meetup:
>
> @derekwaynecarr <https://github.com/derekwaynecarr> added a firm
> requirement for ability to introspect container images
>
> The following questions were addressed:
>
>    1. *Will this provide a view into a container's mount namespace or
>    simply container filesystems?* Only container filesystems since
>    there's no way to support mounting the actual mount namespace as a
>    subdirectory (if this becomes possible we should support it).
>    2. *How should resources be accounted?* There was weak consensus that
>    it's reasonable to expect users to include resources consumed by an kubectl
>    exec in the initial pod sizing.
>    3. *Dangling pod problem with kubectl run --rm?* This would be
>    different from kubectl run in that it would be implemented server-side.
>    That is, the kubelet must ensure there are no dangling containers.
>
> The following open questions were added:
>
>    1. How does this affect the pod lifecycle?
>    2. What are the edge cases of current kubectl exec. Can you re-attach
>    to a running exec? How would this be implemented for an image exec?
>    3. Userns + host bindmounts, depending on how userns is implemented,
>    could be messy (hostmount implies no userns, but userns relabeling might be
>    enabled for the whole pod)
>    4. How does this interact with SELinux?
>
> An alternative proposed during the meeting was to allowing adding a
> container to the podspec. This would be a larger change, but perhaps also
> more worthwhile. It could be simplified by introducing a new type of
> ephemeral container similar to the implementation of init-containers. We
> ran out of time before discussing whether this solution could meet the
> other requirements such as container filesystem access.
>
> It's not clear where to go next, so I'm going to work on addressing some
> of the questions raised by the feedback and try to estimate how much work
> would be required to support modifying the podspec of a running pod.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/35584#issuecomment-257708084>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p1JPTjxL5KHnihv1vPlsC7tVb3TCks5q57KrgaJpZM4KgwPp>
> .
>

timstclair commented on docs/proposals/image-exec.md:
```diff
@@ -0,0 +1,278 @@
+
+We could instead modify CRI's `CreateContainer` to include a list of containers
+IDs to "link" to the newly created container, but then we'd have to repeat the
+logic in each of the container backends.
```
I think this might be unavoidable. If the container runtime uses something other than linux namespaces for isolation (e.g. VM containers), the Kubelet wouldn't be able to manage it.

verb replied:
That's an interesting point. How would pods look under VM containers? Would they still have a shared (e.g. network) namespace somehow? Or would they be pods in name only?

feiskyer replied:
@verb e.g. for hyper, all containers belonging to the same Pod is running at same VM, and they share same network namespace. And different Pods are running at different VMs.

timstclair commented on docs/proposals/image-exec.md:
```diff
@@ -0,0 +1,278 @@
+there are several edge cases to consider such as resource limits and monitoring.
+Adding a volume to a running container would be a smaller change, but it would
+still be larger than the proposed change and we'd have to figure out how to
+build and distribute a volume of binaries.
```
Related: https://github.com/kubernetes/kubernetes/issues/831

timstclair reviewed (commented):
> There was weak consensus that it's reasonable to expect users to include resources consumed by an kubectl exec in the initial pod sizing.

This might be acceptable for an initial implementation, but I think it's far from ideal. If you're running a very large cluster, e.g. 50,000 pods, even a small constant overhead will add up very quickly. It also requires a lot of foresight, which can't always be counted on when debugging something that failed.

verb commented:
@smarterclayton Who could I contact to ask about Huawei's and Red Hat's use cases?

verb commented:
Status update: I was able to prototype adding a container to the spec of a running pod with surprisingly little effort. I need to think through the consequences and edge cases, but I'm optimistic this might be a cleaner solution.

Some thoughts regarding the problem I'm trying to solve:

- Being able to attach/detach from the debugging shell is very nice
- the `kubectl patch` incantation required to do this isn't particularly user friendly. I'd really want there to be an easier command.
- This doesn't address the container filesystems, but that could potentially be added as a general option for a Container
- For troubleshooting pods all I need is support for adding a container (i.e. it doesn't have to be ephemeral) but other use cases may require a way for the container to exit short of deleting the pod.

liggitt commented:
The security and scheduling implications of making the pod spec mutable are concerning. Much has been built on the write-once nature of the security context, volume, port, and resource aspects of the pod spec.

smarterclayton commented:
Not being able to clean up terminated pods immediately will have lots of
implications for cluster stability (for instance, we'll have to detach
shared volumes almost immediately for use cases like jobs that reuse
those).  It's possible that the attached container could hold the pod from
going to terminated state, but that has backwards compatibility concerns
for clients.

On Thu, Nov 3, 2016 at 9:26 PM, Jordan Liggitt <notifications@github.com>
wrote:

> The security and scheduling implications of making the pod spec mutable
> are concerning. Much has been built on the write-once nature of the
> security context, volume, port, and resource aspects of the pod spec.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/35584#issuecomment-258322059>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p7o4FJ0kWpZUhPsNofuw-DsgNkDGks5q6onXgaJpZM4KgwPp>
> .
>

smarterclayton commented:
Having to manage exec sessions via the API has been something I've been
dreading (Docker ended up having to do it).  I think it's likely we'll need
to have the concept of "running sessions" whether it's exec or these new
type of containers, so it's worth asking how much complexity we're going to
add there (maybe we should consider these containers more like exec
sessions, and just have the ability to define an exec environment that is
either in an existing container or creates a new temporary container).  But
that still doesn't remove the need to know that they are there.

On Thu, Nov 3, 2016 at 10:00 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Not being able to clean up terminated pods immediately will have lots of
> implications for cluster stability (for instance, we'll have to detach
> shared volumes almost immediately for use cases like jobs that reuse
> those).  It's possible that the attached container could hold the pod from
> going to terminated state, but that has backwards compatibility concerns
> for clients.
>
> On Thu, Nov 3, 2016 at 9:26 PM, Jordan Liggitt <notifications@github.com>
> wrote:
>
>> The security and scheduling implications of making the pod spec mutable
>> are concerning. Much has been built on the write-once nature of the
>> security context, volume, port, and resource aspects of the pod spec.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/35584#issuecomment-258322059>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABG_p7o4FJ0kWpZUhPsNofuw-DsgNkDGks5q6onXgaJpZM4KgwPp>
>> .
>>
>
>

resouer commented on docs/proposals/image-exec.md:
```diff
@@ -0,0 +1,278 @@
+runtimes.  We could implement the same create/start/attach/remove logic in rkt
+and docker_manager, or we could save effort by linking this feature to the CRI
+and return "Not Implemented" for the legacy runtimes.
+
```
@verb Could you please add a note for hypervisor based runtime e.g.[https://github.com/hyperhq/hyperd] here or other place? Generally we think this will not break hypervisor based runtime (though may require refactoring on runtime side), but we should definitely keep them in mind.

verb commented:
@smarterclayton's idea of exec sessions would address a lot of the usability concerns I have with modifying the podspec as a troubleshooting solution. I'm going to look a bit more into pod/container lifecycles then update the doc.

verb committed:
Rewrite image-exec to broader pod troubleshooting proposal

verb commented:
I've just pushed a pretty significant rewrite of this proposal with a new approach that builds upon existing feature requests. It's larger in scope and will take longer to implement, but it would result in changes that would be generally useful outside of my intended use case.

I think this version addresses most of the feedback so far, but I don't know if it will be more or less popular. Feel free to state a preference for the prior approach.

Note: Proposals have moved to the community repo, but it was suggested to continue discussion in existing PRs so as not to lose context.

smarterclayton commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,317 @@
+
+#### Shared PID Namespace
+
+Tracked in [#1615](https://issues.k8s.io/1615), a shared PID namespace became
```
How complex is this in CRI - i.e. does this require additional changes to CRI interfaces beyond the obvious?

philips replied:
I do not believe it does. But, cc @yujuhong.

feiskyer replied:
CRI could be changed to support this. But I'm concerned not all runtimes supporting shared PID namespace. cc @euank @mrunalp 

Is shared PID namespace a must-have for debugging?

philips replied:
@feiskyer I think shared namespaces are pretty critical. The most common things will be strace and gdb.

verb replied:
It's a must have-have for debugging, and it's something we should support regardless of its use in debugging. (e.g. for a sidecar signalling a main process)

I expect it would only require a change in the CRI if it was optional. @jonboulle suggested in https://github.com/kubernetes/community/pull/207 that the CRI simply mandate a shared PID namespace where possible, which seems consistent with the pod abstraction and worth considering.

I always assumed the only reason we ever had an isolated namespaces were due to limitations in docker. We could simplify several things if the shared namespace were enabled whenever possible.

smarterclayton commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,317 @@
+
+The container will continue to be part of the pod spec, allowing automation to
+inspect the pod, but its resources could be freed by the container runtime. The
+container could be started again manually by performing an update on the pod
```
Adding a net new container, removing then adding the container, or changing the container spec somehow?  Clarify.

smarterclayton commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,317 @@
+this use case (and others?). This would be similar to things like `oc debug` and
+could fit into longer term troubleshooting strategy, but it's not technically
+necessary. A new command might be less confusing for someone expecting `kubectl
+exec` to mirror `docker exec` exactly.
```
This alternative requires 0% new conceptual changes to pods which I think is worth pointing out in the proposal.  "Adding a container to pod spec" and "Container runOnce" are huge changes to the system and thus have to meet a proportionally higher bar of change.

Can you articulate other use cases for "container runOnce" and "adding a container to pod spec" beyond debugging that justify the expense of that change?  How could "container runOnce" be abused?  I consider that within the scope of this proposal to answer and defend against those points.

philips replied:
@smarterclayton how is oc debug implemented from a high level?

smarterclayton commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,317 @@
+all modification to a pod spec as a conservative default and because arbitrary
+changes could have complex repercussions on resource accounting and quality of
+service. The most straightforward way to add a container to a running pod is to
+modify the pod spec.
```
How will this be made backwards compatible / tolerate older clients and tools that assume pod containers are immutable?  What are the risks of introducing this for clusters that have already implemented API tools on top of the platform that assume that pod containers is immutable?  How do we help transition users across that?

smarterclayton replied:
Nm, you answered that last bit.  So just the first questions.

philips replied:
@smarterclayton Are there particular older clients and tools you are concerned about assuming immutability of the spec? I assume that is what you are referring to since a Pod is somewhat dynamic in that containers could be restarted, etc changing the status.

My understanding is that part of the motivation for the CRI is to make Pods a scheduling unit but explicitly something that can be mutated later.

smarterclayton replied:
Primarily anything security related around pods, because their immutability today makes enforcement easier than otherwise.  I can think of 5-10 protections that depend on it.  I agree we will relax mutability, but at this point we need to carefully assess each change and make sure we're adding other controls to allow it (like field level auth)

philips replied:
What are those 5-10 protections?

verb replied:
I added a section to discuss necessary updates to external tools.

smarterclayton commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,317 @@
+override the pod's `restartPolicy` for the newly added container so it is not
+restarted when it exits. When a "run once" container exits, it will transition
+to the Terminated state, not be restarted, and continue to assert the `Ready`
+condition.
```
How can runOnce containers be abused / used outside of debugging?  More details.

smarterclayton commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,317 @@
+QoS class could drop from *Guaranteed* to *Burstable*. This is an acceptable
+consequence of modifying a running pod. It means a pod could potentially be
+evicted during debugging, particularly if there are insufficient resources to
+allow for the troubleshooting process.
```
Which means practically any resource limited / tuned job needs extra resources allocated.  I guess link to vertical resizing proposal here (or at least call it out).

derekwaynecarr replied:
i am still not a fan of letting a running pod's qos class change.

verb replied:
Based on my quick reading of pod cgroup the qos class won't change. I've added pod cgroup as a prereq and made a note to confirm.

verb replied:
I called out vertical resizing, but I couldn't find a proposal (and I didn't know about it prior to your comment).

smarterclayton commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,317 @@
+Admission controllers will watch for this operation in order to approve the
+images from the modified pod spec. Containers can be added regardless of pod
+status to allow for troubleshooting containers in states like
+`CrashLoopBackoff`.
```
Special note - adding a container requires that pod security policy pass for the person doing the adding - which may not be possible in all cases (the entity creating the pod has higher on node privilege than the user, which means the user is simply not allowed to access the container). Since pods are atomic today we do container resolution for one user - this addition is like exec and we'll likely need to apply the same rules.

Adding a container is the equivalent check to "can a user add all the existing pods" since the added container will be able to act on the node as any of the existing containers.

smarterclayton commented:
Barring the complexity and number of prerequisites required to bring this to completion (and i still believe debug is an 80% solution that we should do before this), i do think it bounds the challenges we had before.  Allowing only container addition is pretty important, as is determining how susceptible
container runOnce is to abuse / misuse.

verb commented:
@smarterclayton That's a really great point. I'll update the doc to propose the 80% solution first thing, which will be more effort but with immediate benefit. The debug solution can gain functionality as the prerequisites fall in to place, ending with the ability to "attach" to a running pod (eventually). Thanks for the quick feedback!

yifan-gu commented:
cc @lucab @s-urbaniak @euank

philips commented on docs/proposals/pod-troubleshooting.md:
````diff
@@ -0,0 +1,317 @@
+% for pod in $(kubectl get -o name pod); do
+    kubectl exec -m gcr.io/neato/security-audit $pod /security-audit.sh
+  done
+```
````
On every run this leaves garbage around in the runOnce case though, right?

verb replied:
I've switched to tracking these in a new field so this use case is better supported.

philips commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,317 @@
+In order to support short-lived processes (e.g. troubleshooting sessions,
+automation) from the general pod spec container list we need to be able to
+override the pod's `restartPolicy` for the newly added container so it is not
+restarted when it exits. When a "run once" container exits, it will transition
```
Doesn't this leave garbage around, particularly if there is automation on top?

verb replied:
Yes, it does by design in order to meet the requirement that changes can be audited by existing tools. Alternatively we could garbage collect and surface the changes in some other field of Pod/PodStatus, but tools would have to be updated to check a new field.

I think @derekwaynecarr added this requirement, but I may have misunderstood it.

philips commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,317 @@
+An earlier version of this proposal suggested running an ephemeral container in
+the pod namespaces. The container would not be added to the pod spec and would
+exist only as long as the process it ran. This has the advantage of behaving
+similarly to the current kubectl exec, but it is opaque and likely violates
```
It also has the advantage of not cluttering up the spec with "dead" containers, right?

verb replied:
I've moved the reporting of these containers to a new field in PodStatus so the container spec won't become crufty.

philips commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,317 @@
+
+This would be a small change to Kubernetes, but it would create edge cases in
+the pod lifecycle that would have to be considered. For example, what happens to
+the debugging pod when the target pod is destroyed?
```
Presumably it would be destroyed as well as it is a child of the namespaces. 

I do like this approach because the debug pod may be resource intensive and it gets accounted differently than the debug target.

verb replied:
Should the new pod evict other pods if the node has insufficient resources? I'm not sure of the current behavior when a podspec specifies a nodeName.

I lean towards requiring the troubleshooting to fit within the existing pod resources, especially if we were to give the user a choice of either "attaching" to the running pod or creating a new copy of the pod with increased resource requirements.

vishh replied:
+1 for this approach.

> Should the new pod evict other pods if the node has insufficient resources? I'm not sure of the current behavior when a podspec specifies a nodeName.

That depends on the priority of the debug pod. I suspect that users will not want to affect existing critical pods and so debug sessions might run with minimal resource requirements. I'd recommend running them as `Burstable` pods to have at-least minimal resource guarantees. 

> I lean towards requiring the troubleshooting to fit within the existing pod resources, especially if we were to give the user a choice of either "attaching" to the running pod or creating a new copy of the pod with increased resource requirements.

There is no notion of `pod resources` in k8s. What you are proposing is using `slack resources` based on the sum of all container's resources. We plan on using this `slack` for multiple purposes. On one extreme a debug container might not even start because of lack of memory for example or CPU. This is highly un-desirable since you cannot `debug`. 

I suspect that the uncertainty with injecting a container into a pod spec will make the feature brittle.

philips reviewed (commented):
I agree with the overall goals of this proposal. I think there is some more noodling that needs to happen around the technical approach.

derekwaynecarr commented:
Just capturing the question I had from sig-node around the use case of debug a copy of a pod is that copy does not land on the same node as the original source pod.  For example, I assume the copy needs to have access to all the source pods volumes.

verb committed:
Update Troubleshooting Proposal for Debug Command

This update encapsulates a longer-term troubleshooting vision that
includes a new command `kubectl debug` that allows troubleshooting a
copy of a pod in the short-term and the running pod in the long-term.

verb committed:
Add examples to pod-troubleshooting proposal

Also fixes typos and integrates early feedback. Open Questions move to
github issue.

verb commented:
Further iteration on design based on feedback. Highlights:

- Addition of a `kubectl debug` command to become the user's interface to this functionality.
- A *Copy Mode* that can be implemented in the existing API (@smarterclayton's 80% solution)
- Pod spec is immutable, new container is attached by kubelet and reported in PodStatus (@dchen1107's suggestion)

I'll provide an update at the next SIG Node meeting.

thockin commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,489 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
```
please remove the UNVERSIONED_WARNING section - it's obsolete

thockin commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,489 @@
+kubectl debug -p example -c shell --image=debian
+# Container DNE, create a copy with an additional container
+kubectl debug -p example-copy --copy-of example -c shell --image=debian
+# Container name exists, create a copy with a different entrypoint
```
I'm less excited by this case

thockin commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,489 @@
+The status of a *debug container* is reported in a new `DebugContainerStatuses`
+field of `PodStatus`, which is a read-only list of `ContainerStatus` that
+contains an entry for every *debug container* that has ever run in this pod.
+Additionally, a new boolean `Tainted` in `PodCondition` changes from false to
```
This is sort of at the edge of status's intended meaning.  The litmust test was always "if status were lost, could it be recreated by inspection" ?  I guess it could in this case, if the runtime implementations are careful to squirrel away data in tings like docker labels

verb replied:
Good point. Tainted can be calculated by len(DebugContainerStatuses) > 0 for the case of debug containers but we would have to track additional state for something like kubectl exec so it might not be worth doing. There's no requirement for this, it just seemed like something people might want.

thockin commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,489 @@
+A requirement for prior configuration makes it feel like a workaround rather
+than a feature of the platform.
+
+### Implicit Empty Volume
```
I still think I like this better, and I'm not sure I buy the complexity argument.  Running a new container in the pod has a lot of complexity too.

This, OTOH, feels more natural.  The implicit volume can be dynamically named, so no collisions and can even be read-only to the container.  Debugging then becomes a) provide the name of a container to unpack and b) exec.

I guess it doesn't help much if your main `command` or `args` are busted and you are crashlooping

 Also need access to the root dir of the container's main process, but that seems tractable.

verb replied:
I definitely want the thing that feels most natural. For me that's running a container since that's what we do everywhere else in Kubernetes. We by necessity have an infrastructure for distributing container images so I'd prefer to use that rather than have the kubelet/runtime unpacking tarballs, though it might be possible to do something like union mount a container image at debug time.

Adding a container doesn't require any additional complexity on the run times, which is nice given the diversity we want to support.

Making this work with the existing `kubectl exec` (by adding binaries) makes a lot of other things easy, but I don't like surfacing that complexity to the user. Let's say we mount debian's userland at /adc2ce42-e1af-11e6-bf01-fe55135034f3. kubectl can figure out how to exec /bin/bash by looking up its mount point, but bash has no way to figure out that curl is at /adc2ce42-e1af-11e6-bf01-fe55135034f3/usr/bin. We can't set env variables for exec, so we need to embed a launcher within the image. kubectl calls `/adc2ce42-e1af-11e6-bf01-fe55135034f3/launch bash` which sets PATH=/adc2ce42-e1af-11e6-bf01-fe55135034f3/bin and exec()s bash. If the user wants to know what utilities are installed, `ls /bin` will be very confusing. Instead the user should `which bash` or `ls /` and then `ls /adc2ce42-e1af-11e6-bf01-fe55135034f3`. If a tool is missing, `apt-get` won't work, you have to mount a different image, which might unpack tools over the top of your mount point.

The thing I like about running a new container is that it behaves like a regular Linux userland.

thockin reviewed (commented):
I don't want to derail this, so if I am outgunned, I will cope with it.

thockin commented:
> Making this work with the existing kubectl exec (by adding binaries) makes a lot of other things easy, but I don't like surfacing that complexity to the user.

Nothing should surface to user, or it is a non-starter.

> Let's say we mount debian's userland at /adc2ce42-e1af-11e6-bf01-fe55135034f3. kubectl can figure out how to exec /bin/bash by looking up its mount point, but bash has no way to figure out that curl is at /adc2ce42-e1af-11e6-bf01-fe55135034f3/usr/bin. We can't set env variables for exec, so we need to embed a launcher within the image.

I assumed it would run as chroot, just like a container, but with the
container's rootfs mounted via bind mount or something.  This way you
get the benefits of beng a container without the weirdness of not
literally being in the same container to debug.

Admittedly, if we have shared PIDNS and we can find a way to
cross-mount container rootfs's and volumes, it's pretty much moot.

octo commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,489 @@
+pod "example" created
+
+# Container "shell" does not exist, add a new container to running pod
+kubectl debug -p example -c shell --image=debian
```
I would like to propose to change the syntax to `kubectl debug POD [args …]` where `POD` is *always* the pod you'd like to debug. The current proposal has the confusing semantics that the pod you want to debug is sometimes passed as `-p POD` and sometimes as `--copy-of POD`.

I'll try to attempt to propose an alternative syntax by example:

```
# Container "shell" does not exist, add a new container to running pod
# N.B.: --in-place is borrowed from GNU's sed.
kubectl debug example --in-place -c shell --image=debian 

# Container DNE, create a copy with an additional container
# If -p/--pod is not given, a name is automatically derived.
kubectl debug example [-p example-copy] -c shell --image=debian

# Container name exists, create a copy with a different entrypoint 
kubectl debug example [-p example-copy] -c container-with-distro --command -- /bin/sh

# ERROR: cannot modify entrypoint of running pod (pod spec is immutable)
# (aka. --in-place and --command are mutually exclusive)
kubectl debug example --in-place -c container-with-distro --command -- /bin/sh # ERROR! 
```

I'd argue this has the following benefits:

*   `kubectl debug POD-TO-DEBUG` is consistent (both, across the different modes and other kubectl subcommands), easy to memorize and intuitive.
*   Inverting the logic of `--copy` to `--in-place` allows us to implement the *copy debug mode* first and add the *running debug mode* later without clumsy defaults or break the CLI.

    With `--copy`, we'd either
    *   have to require people type `--copy` always (even though the flag wouldn't serve a purpose at first) and make *running debug mode* the default when it's available,
    *   end up with `--nocopy` which is not very descriptive because is disables implicit behavior, or
    *  change the default behavior once *running debug mode* is implemented.

I'm not particularly happy with specifying the name of the cloned pod via `-p CLONED-POD` but I fail to come up with a good alternative. Thoughts?

verb replied:
I agree with your analysis and like your suggestions better than the original. I especially like having a consistent debug "target" specified as an argument rather than a flag.

One correction for your example, it's not `--in-place` and `--command` that are mutually exclusive, it's that `--in-place` cannot be used with a container name that's already in the pod spec. Also, I agree that using `-p/--pod` to refer to a pod that doesn't exist is inconsistent with the rest of `kubectl`. Maybe `--copy-name`?

One nit is that `--in-place` might imply one can edit the pod spec in place, which isn't true, but I can't think of a better way to convey the semantics. Just brainstorming some things:

```
# ERROR: container  'nginx' already exists
kubectl debug nginx --in-place -c nginx --command -- /bin/sh

# Running debug adding a new container with generated defaults for everything, including image.
# Is a default image a bad idea?
kubectl debug nginx --in-place

# If specifying an image is required then we could make that the argument that activates running debug
kubectl debug nginx --run busybox
# ERROR: awkward side effect of above
kubectl debug nginx --run busybox --image busybox

# What would we want this to do?
kubectl debug nginx
```

smarterclayton replied:
Specifying a new pod name to debug is an edge case.

All debugging is about the target, so the argument should be a single value which supports our consistent "acts on pod-like things" - deployments, RS, or pods (with pod being the default resource type).  There will have to be controller specific behavior for many cases, but that's actually desirable (we want stateful set debug not to crash the set).  That also leaves the door open to future semantic options - "--under-service=foo", for example.

A default image is probably not practical for many reasons, starting with library compatibility all the way up to security concerns.  

We might be trying too hard to cram "create a copy to debug" and "add a container" under the same verb.  "Add a container" feels more like an attach variant.

octo replied:
Not all proposed uses *create* a container. For example, you might want to pass an additional `-debug` or `-trace` option to an existing container / process.

I think there are two (three) options for the behavior when `-c` is missing:

*   Use the first container of the pod.
*   Create a new container.
*   (Do one of the above, depending on context. E.g. default to creating a new container when `--image` is passed and the first container otherwise. I'd call this the "Perl approach" ;)

I have the impression that the first option would fit in best with other subcommands of `kubectl`.

vishh commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,489 @@
+make the pod spec mutable so the client can generate an update adding a
+container. `SyncPod()` has no issues adding the container to the pod at that
+point, but an immutable pod spec has been a basic assumption in Kubernetes thus
+far and changing it carries risk. It's preferable to keep the pod spec immutable
```
This will change with `vertical pod auto-scaling` for example. Before we preclude pod spec updates, can we at-least list the issues with mutating pod spec?

verb replied:
I think sig-node's objection to mutable pod spec is a philosophical one that a pod/task/etc should be immutable as a best practice. For example, mutable pods could lead to people creating and relying on indefinitely-lived pods, and that's now how kubernetes was designed.

As far as I know there aren't any technical roadblocks.

vishh commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,489 @@
+attach a volume at a random path at run time, just in case it's needed. Though
+this simplifies the solution by working within the existing constraints of
+`kubectl exec`, it has a sufficient list of minor limitations (detailed in
+[#10834](https://issues.k8s.io/10834)) to result in a poor user experience.
```
I'm posting an alternate workflow for debug volumes here just to facilitate a discussion. It is possible that you have already considered this idea and I have failed to identify the disadvantages with such an approach in your comments on #10834. Kindly bear with me.

1. It is possible to export a docker image into a volume using `docker export`
2. Such a volume can be injected into a container at runtime using mount propagation
3. An exec session in a container can then use the tools within the new debug volume. To work around PATH issues, the exec session can `chroot` assuming `chroot` is available in the debug volume. To have access to the container's root filesystem, the container's root filesystem can be mounted into the volume at a special path.
4. Once the special exec session ends, the debug volume can also be cleaned up.

vishh reviewed (commented):
I posted some comments on the primary design idea.
I'm still not convinced that the alternate design ideas you have included are not viable.

octo commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,489 @@
+`kubectl exec`'s "execute a debug process in a container" in a new command,
+`kubectl debug`, which has two modes of operation:
+
+1.  *Copy Debug* mode creates a copy of a pod with minor changes to the pod spec
```
I think the doc describes use cases in sufficient detail. What are your reasons for not wanting to add this feature as described? It's hard to argue with feelings …

That said, I tend to agree that *debug* is not the best subcommand / verb for the *Copy Debug* mode because it is not discoverable / intuitive: reading through the list of available subcommands I would not expect it to do what it does. *kubectl clone pod* might be better in that respect and could eventually be extended to work on other types of resources as well.

verb commented:
Status update: I have a pretty decent working prototype now and am working on an implementation plan. I think SIG CLI wants to break "Copy Debug" mode (a.k.a. make a new copy of a running pod with some config changes) into a separate feature and implement it elsewhere. As soon as I get things ordered I'll update this proposal.

verb committed:
Refine Pod Troubleshooting proposal

* Add additional Use Cases
* Simplify Copy Debug vs Running Debug distinction
* Add requirements analysis

verb committed:
Update Pod Troubleshooting proposal with 1.7 alpha goals

verb commented:
/release-note-none

@k8s-bot node e2e test this

(doesn't matter sense this will be closed here and merged into the community repo, but it takes the red X off my dashboard)

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+
+*Running Debug* requires new functionality in the kubelet and a new `kubectl
+debug` command that results in a new container being introduced in the running
+pod. This *debug container* is not part of the pod spec, which remains
```
Without modifying the pod spec, how does kubelet know to create the new debug container?

yujuhong replied:
I think the doc evolves over time and the result is that many important points are not mentioned early enough. I'd suggest giving the overview of the mechanism before diving into specific details.

verb replied:
You're right, I think a rewrite may be in order.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
````diff
@@ -0,0 +1,573 @@
+```
+
+If the specified container name already exists, `kubectl debug` will kill the
+existing container prior to starting a new one. It's legal to reuse container
````
Why not error out and let users decide whether to kill the previous debug container or not? Replacing the container directly may cause confusion.

yujuhong replied:
I think asking the user before replacing it is more reasonable. 

Another thing is that the container names cannot be reused as long as the container still exists, even if it already exited. You'll need to remove (GC) the container, which may cause the loss of one "DebugContainerStatus".

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,489 @@
+lack of configuration and restart policy. The kubelet's additional
+responsibilities include:
+
+1.  Start or restart a debug container in a pod when requested.
```
Without modifying the pod spec, you'd be relying on kubelet to *remember* this request. If kubelet restarts before creating the container (or if the container crashes for other reasons), it will never be (re-)created without kubelet persisting the request somewhere. This *may* be okay but it should be stated clearly in this proposal.

yujuhong replied:
I'd suggest you move the "exec" analogy to the beginning of the section. I think that'd help people understand the proposed method much better.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+pod. This *debug container* is not part of the pod spec, which remains
+immutable, and is not restarted according to the pod's `restartPolicy`.
+
+The status of a *debug container* is reported in a new `DebugContainerStatuses`
```
What does `DebugContainerStatuses` include? If the "name" of the debug container can be reused, how do users distinguish different instances of the debug containers? Will there be an ID assigned to the instance, or will you just rely on the "container ID" reported by the runtime?

yujuhong replied:
If there are multiple active debug containers, how could the user tell which one is created by them? Do they rely on the "name" of the containers to figure that out? If that's the case, since you will not GC the terminated debug containers, users can never reuse the same name. I think that's ok, but this is not explained clearly in the proposal.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+1.  Don't kill the debug container while the pod is alive.
+1.  Report status on the debug container.
+1.  Stop and delete the container when the pod is deleted.
+1.  Preserve and report state of exitted debug containers
```
s/exitted/exited

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+Of these, preventing `SyncPod()` from killing the container is the most complex,
+but most of this complexity has already been implemented by *init containers*.
+For *debug containers* we need only amend `computePodContainerChanges()` to
+ignore containers labeled as debug.
```
If you have debug containers with different names, how does kubelet know which one to keep? Do you allow multiple debug containers to co-exist? If not, do all instances reuse the same name?

verb replied:
Yes, multiple debug containers are allowed to co-exist. All will be kept, but they'll all be in the same cgroup which could lead to the pod being terminated for exceeding its resources.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+1.  Preserve and report state of exitted debug containers
+
+Of these, preventing `SyncPod()` from killing the container is the most complex,
+but most of this complexity has already been implemented by *init containers*.
```
I'm not sure if init containers help a lot here -- init containers are included in the pod spec and are closer inherently to how kubelet examine regular containers.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+
+The simplest way to preserve state of debug containers is to exempt them from
+garbage collection while the pod exists. Number of debug containers are limited
+because they are started manually and are never restarted.
```
It should be noted that this may cause resource problems on the node, causing pods to be evicted. I think it's a reasonable tradeoff to delete debug container instances when the node is under resource pressure.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+
+To distinguish debug containers, the runtime uses a new
+`io.kubernetes.container.type` label. Existing containers will be started with a
+type of `REGULAR`, `INIT` or `SANDBOX`. When added in a subsequent step, debug
```
Whether the container is acting as a pod sandbox or not is an implementation detail of the runtime. Kubelet does not care about it. It should not be included in this label.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+runtime via labels.
+
+To distinguish debug containers, the runtime uses a new
+`io.kubernetes.container.type` label. Existing containers will be started with a
```
The code needs to be backward compatible to handle containers created by older kubelets.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+##### Step 1: Container Type
+
+Debug containers exist within the kubelet as a `ContainerStatus` without a
+container spec. As with other `kubecontainers`, their state is persisted in the
```
s/state/type ?

yujuhong replied:
I wouldn't say we persist the state (which changes). We mostly persist the spec.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+This step adds methods for creating debug containers, but doesn't yet modify the
+kubelet API. The kubelet will gain a `RunDebugContainer()` method which accepts
+a `v1.Container` and creates a debug container. Fields that don't make sense in
+the context of a debug container (e.g. probes, lifecycle, resources) will be
```
Please define the valid fields for a debug container.

verb replied:
Done

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+
+The kubelet will treat `DEBUG` containers differently in the following ways:
+
+1.  `SyncPod()` ignores containers of type `DEBUG`, since there is no
```
What happens if the pod restarts (e.g., infra container died and gets recreated)? What would happen to the debug container?

yujuhong replied:
But the proposal said `SyncPod` would just ignore the debug containers.... It can't really ignore AND compute changes for the debug containers. I'm a little bit concerned about the increased complexity of the code which may *at times* ignore the debug container. I know this is the implementation detail, but please make sure the logic clear and readable in the code.

verb replied:
You're right, there's an edge case in sandbox restarts where a debug container should be killed prior to pod deletion, but it's not a complex one. I've updated the proposal to reflect this. If you'd like to see the added SyncPod complexity, it's here: 

https://github.com/verb/kubernetes/pull/3/files#diff-8da51ab012ecbdf2b013ecbcf1c37033

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+
+##### Step 3: kubelet API changes
+
+The kubelet will gain a new streaming endpoint `/debug` similar to `/exec`,
```
kubelet already has a `/debug` endpoint. Need another name.

verb replied:
whoa, somehow missed that. Ok, will rename to.. umm... "podDebug", which seems similar to existing endpoint "containerLogs"

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+
+The kubelet will gain a new streaming endpoint `/debug` similar to `/exec`,
+`/attach`, etc. The debug endpoint will create a new debug container and
+automatically attach to it.
```
Is it allowed to create a non-interactive debug container?

Also, creating and starting a container may take a long time (e.g., pulling down the image). Wouldn't this affect the user experience?

verb replied:
I'd like to allow for the non-interactive use case, but not necessarily the detached use case.

If it took a very long time container creation might affect user experience, but I don't see any way around it. definitely open to suggestions.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,489 @@
+responsibilities include:
+
+1.  Start or restart a debug container in a pod when requested.
+1.  Don't kill the debug container while the pod is alive.
```
@verb IIUC, since you treat the 'debug' session similar to the 'exec' session, the debug container will terminate whenever the command finishes, or the user disconnects? In other words, there should not be a running debug container with an active user streaming session. Is that correct?

verb replied:
I went ahead and added reattaching debug containers to the proposal, because it's easy and I think it's useful.

yujuhong commented on docs/proposals/pod-troubleshooting.md:
```diff
@@ -0,0 +1,573 @@
+
+The status of a *debug container* is reported in a new `DebugContainerStatuses`
+field of `PodStatus`, which is a read-only list of `ContainerStatus` that
+contains an entry for every *debug container* that has ever run in this pod, and
```
I think keeping a maximum "terminated" debug containers is an acceptable first step.

By the way, what's the purpose of keeping a list of ever-growing debug container statuses? We will have auditing soon for posterity already.

verb commented:
/unassign @smarterclayton
/unassign @thockin 
/assign @yujuhong 
/assign @dashpole

verb committed:
Rewrite Pod Troubleshooting proposal

* Separate "Copy Mode" to another PR for another time.
* Describe feature in similarity to `kubectl exec`
* Simplify feature description, move non-essential supporting sections
  to Appendices.

verb commented:
Rewrite to implement @yujuhong's suggestions re: format. Simplify and move non-essential info to Appendices. Also addresses all comments from previous commit (afaik).

dashpole commented:
@verb regenerating the podSandboxConfig when creating a debug container should not be problematic, since we regenerate it when restarting containers as well.

dashpole commented on docs/proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,568 @@
+
+This creates an interactive shell in a pod which can examine and signal all
+processes in the pod. It has access to the same network and IPC as processes in
+the pod. It can access the filesystem of other processes by `/proc/$PID/root`,
```
why organize the filesystem by process instead of container?  Does this mean that if I have 2 processes within a container, ill have two different ways of seeing the same root filesystem?  Would it make more sense to place the filesystems as `/container_name/`?

verb replied:
It's just because we get that for free with Linux. Originally I had intended to cross-mount the filesystems as you suggest, but that requires a change to the CRI which was contentious, so we agreed to start with this.

You're correct that there would be multiple ways of viewing the same filesystems, though technically they could be different mount namespaces.

dashpole replied:
Ok, that makes sense.

dashpole replied:
How does this work if there are multiple debug containers in the same pod?  Do the debug containers see each other?

verb replied:
Yes, as long as there's a shared PID namespace, which is enabled by default for 1.7. This feature relies on a shared PID namespace.

dashpole commented on docs/proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,568 @@
+One way in which `kubectl debug` differs from `kubectl exec` is the ability to
+reattach to Debug Container if you've been disconnected. This is accomplished by
+running `kubectl debug -r` with the name of an existing, running Debug
+Container. The `-r` option to `kubectl debug` is translated to a `reattach=True`
```
why not implicitly reconnect, instead of making the user specify it?  The only case I can think of is if the user wasnt aware that the debug container was running, and somehow mucks with a running debug container accidentally.

verb replied:
I like the idea of implicit reconnection, but I'm not sure how to handle the container parameters in that case. For example, say one user runs `kubectl debug -it pod -m debian -c debug -- bash` and, while that's running, another user runs `kubectl debug -it pod -m alpine -c debug -- ls /`. 

The kubelet will receive a v1.Container with an image and args that differs from the existing container. The easiest thing to do would be to discard the new v1.Container and attach to the running container, but that would be confusing to the second user.

We could compare the provided v1.Container to the existing ContainerStatus and only reattach if the run time arguments are identical, but that might still be confusing for user 2 if she didn't notice user 1, so I just opted for requiring explicit behavior.

I'm a big fan of implicit behavior and reasonable defaults, though, so I'd be happy to change this if we can work out the details.

dashpole replied:
That makes sense.

verb replied:
I updated the alpha feature scope section of the doc to reflect this.

dashpole commented on docs/proposals/troubleshoot-running-pods.md:
```diff
@@ -0,0 +1,568 @@
+is an error to attempt to replace a Debug Container that is still running.
+
+One way in which `kubectl debug` differs from `kubectl exec` is the ability to
+reattach to Debug Container if you've been disconnected. This is accomplished by
```
Does the debug container exit when you disconnect from it? That would seem like generally desirable behavior to me, as it would guarantee that only debug containers that actively are being used stick around.

verb replied:
If you're disconnected involuntarily (network interruption, pkill -9 kubectl, etc) then the container continues running so that you can reattach where you left off. The only way to disconnect intentionally is to terminate the process (exit a shell, ^C a process, etc) which would cause the container to exit.

We could improve upon this by adding an escape sequence to kubectl, but probably not in time for an alpha in 1.7.

dashpole replied:
just curious: Is there a use case around debugging with state?  Basically, is there ever a case where a user would care that they start a new container rather than return to an older debug container?

verb replied:
The only person that's mentioned it so far is @smarterclayton here: https://github.com/kubernetes/kubernetes/pull/35584#issuecomment-258328491

I like the idea of being able to reattach, but it's not a hard requirement for our own use case. I kept it for the alpha because it was so easy to implement.

dashpole replied:
Right.  I guess my question is more around whether or not to expose that functionality to users if they dont need it.  With regards to @smarterclayton's comment, if a user needs to be aware of the different debug containers that are active in order to use the tool effectively, then we need to expose that to them.  The alternative is to make the debug command act the same way regardless of what other debug containers exist, similar to how exec works (I think).  Essentially the behavior would be to always start a new debug container, and always clean up the debug container after the session exits.

It isn't necessary to make the decision before the alpha implementation though.  But it may be difficult to remove reattach functionality after we add it, even if it is just an alpha.

verb replied:
No matter what we choose for reattach, active debug containers will be listed with names via `kubectl describe pod`

My gut feeling is that reattach will be something people want, but I don't want to paint us into a corner. What if we just do implicit reattach for the alpha and live with the surprising behavior? That way we haven't added any parameters to the API that we might have to support.

dashpole replied:
I think that would work well.

verb replied:
I updated the alpha feature scope section of the doc to reflect this.

dashpole reviewed (commented):
Implementation looks good.  Just a couple general questions.  Apologies if they have already been discussed

verb committed:
remove garbage collection requirement from alpha

verb commented:
@dashpole Awesome, thanks for checking on podSandboxConfig.

@dashpole and I agreed offline that modifying GC for the alpha feature isn't necessary since it would likely change before graduating alpha, anyway. I've updated the doc to reflect this.

verb committed:
Remove reattach from the alpha feature

dashpole commented:
This proposal LGTM

verb commented:
Moved PR to kubernetes/community#649 to be merged.
